third_party/move/move-compiler-v2/src/file_format_generator/function_generator.rs

```diff
@@ -196,6 +199,403 @@ impl<'a> FunctionGenerator<'a> {
+            struct_env.module_env.get_id(),
+            struct_env.get_id(),
+            para_type_parameters.clone(),
+        );
+        let ref_struct_ty = Type::Reference(ReferenceKind::Immutable, Box::new(struct_ty.clone()));
+        let loc = struct_env.get_loc();
+        let function_indices = r#gen.struct_api_test_variant_index(ctx, &loc, struct_env);
+        let mut fun_count = r#gen.module.function_defs.len();
+        let visibility = struct_env.get_visibility();
+        for (variant, function_index) in function_indices {
+            let def_idx = FunctionDefinitionIndex::new(ctx.checked_bound(
+                &loc,
+                fun_count,
+                MAX_FUNCTION_DEF_COUNT,
+                "defined function",
+            ));
+            r#gen
+                .source_map
+                .add_top_level_function_mapping(def_idx, ctx.env.to_ir_loc(&loc), false)
+                .expect(SOURCE_MAP_OK);
+            for TypeParameter(name, _, loc) in struct_env.get_type_parameters() {
+                r#gen
+                    .source_map
+                    .add_function_type_parameter_mapping(
+                        def_idx,
+                        ctx.source_name(name, loc.clone()),
+                    )
+                    .expect(SOURCE_MAP_OK)
+            }
+            let input_para_name = struct_env.env().symbol_pool().make(well_known::PARAM_NAME);
+            r#gen
+                .source_map
+                .add_parameter_mapping(def_idx, ctx.source_name(input_para_name, loc.clone()))
+                .expect(SOURCE_MAP_OK);
+
+            // Function body:
+            // MoveLoc(0)
+            // TestVariant/TestVariantGeneric
+            // Ret
+            let locals = r#gen.signature(ctx, &loc, vec![ref_struct_ty.clone()]);
+            let mut code = vec![];
+            code.push(FF::Bytecode::MoveLoc(0 as FF::LocalIndex));
+            let test_variant_bc = if para_type_parameters.is_empty() {
+                let idx = r#gen.struct_variant_index(ctx, &loc, struct_env, variant);
+                FF::Bytecode::TestVariant(idx)
+            } else {
+                let idx = r#gen.struct_variant_inst_index(
+                    ctx,
+                    &loc,
+                    struct_env,
+                    variant,
+                    para_type_parameters.to_vec(),
+                );
+                FF::Bytecode::TestVariantGeneric(idx)
+            };
+            code.push(test_variant_bc);
+            code.push(FF::Bytecode::Ret);
+            let code = FF::CodeUnit { locals, code };
+
+            let def = FF::FunctionDefinition {
+                function: function_index,
+                visibility,
+                is_entry: false,
+                acquires_global_resources: vec![],
+                code: Some(code),
+            };
+
+            r#gen.module.function_defs.push(def);
+            fun_count += 1;
+        }
+    }
+
+    /// Generates pack api
```

Comment:
nit: it would be nice to have a example API names and signatures generated, eg., for ```public struct S { x: u8 }```, we generate `public fun pack_mangled(_x: u8): S`.